tracing/middleware.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2022 CloudWeGo Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracing
+
+import (
+	"context"
+	"time"
+
+	"github.com/bytedance/gopkg/cloud/metainfo"
+	"github.com/cloudwego/kitex/pkg/endpoint"
+	"github.com/cloudwego/kitex/pkg/klog"
+	"github.com/cloudwego/kitex/pkg/rpcinfo"
+	"github.com/kitex-contrib/obs-opentelemetry/tracing/internal"
+	"go.opentelemetry.io/otel/baggage"
+	"go.opentelemetry.io/otel/sdk/trace"
+	oteltrace "go.opentelemetry.io/otel/trace"
+)
+
+// ClientMiddleware inject span context into req meta
+func ClientMiddleware(cfg *config) endpoint.Middleware {
+	return func(next endpoint.Endpoint) endpoint.Endpoint {
+		return func(ctx context.Context, req, resp interface{}) (err error) {
+			span := oteltrace.SpanFromContext(ctx)
+			if !span.IsRecording() {
+				return next(ctx, req, resp)
+			}
+
+			// inject client service resource attributes (canonical service) to baggage
+			readOnlySpan := span.(trace.ReadOnlySpan)
+			bags, err := injectCanonicalServiceToBaggage(readOnlySpan.Resource().Attributes())
```

Comment:
I think it would be better to propagate `CanonicalService` info by metainfo, not baggage.  Because baggage is designed to propagate value through entire call chain(or you have to then from baggage, it's cumbersome)

Reply:
> I think it would be better to propagate `CanonicalService` info by metainfo, not baggage. Because baggage is designed to propagate value through entire call chain(or you have to remove then from baggage, it's cumbersome)

Good idea, I didn't use metainfo for transparent transmission before because metainfo is a unique feature of kitex

But baggage itself is also implemented based on metainfo, and it will be more concise to use metainfo directly here